File: BeatUpClient/cs/Patches/1_MainSystemInit.cs
static partial class BeatUpClient {
	[Detour(typeof(BeatSaberInit), nameof(BeatSaberInit.InstallBindings))]
	static void BeatSaberInit_InstallBindings(BeatSaberInit self) {
		Base(self);
		Zenject.SceneContext.ExtraPostInstallMethod += container => {
			Injected<GameScenesManager>.Resolve(Zenject.ProjectContext.Instance.Container);
			BeatSaberMultiplayerSessionManager? multiplayerSessionManager = Injected<BeatSaberMultiplayerSessionManager>.Resolve(container);
			IMenuRpcManager menuRpcManager = Injected<IMenuRpcManager>.Resolve(container)!;
			Injected<AudioClipAsyncLoader>.Resolve(container);
			Injected<BeatmapDataLoader>.Resolve(container);
			Injected<BeatmapLevelsModel>.Resolve(container);
			Injected<BGNet.Core.GameLift.GameLiftPlayerSessionProvider>.Resolve<BGNet.Core.GameLift.IGameLiftPlayerSessionProvider>(container);
			Injected<CustomNetworkConfig>.Resolve<INetworkConfig>(container);
			Injected<IMultiplayerStatusModel>.Resolve(container);
			Injected<IQuickPlaySetupModel>.Resolve(container);
			Injected<MultiplayerLevelScenesTransitionSetupData>.Resolve(container);
			multiplayerSessionManager?.SetLocalPlayerState("modded", true);
			Net.Setup(menuRpcManager, multiplayerSessionManager);
		};
	}

	static NetworkConfigSO networkConfigSO = null!;
	[Detour(typeof(MainSettingsAsyncLoader), nameof(MainSettingsAsyncLoader.RegisterInstallers))]
	static void MainSettingsAsyncLoader_RegisterInstallers(MainSettingsAsyncLoader self, BGLib.AppFlow.Initialization.IInstallerRegistry registry) {
		Base(self, registry);
		networkConfigSO = self._networkConfig;
		NetworkConfigSetup(self._networkConfig);
	}

	[Detour(typeof(MainSettingsAsyncLoader.MainSettingsInstaller), nameof(MainSettingsAsyncLoader.MainSettingsInstaller.InstallBindings))]
	static void MainSettingsInstaller_InstallBindings(MainSettingsAsyncLoader.MainSettingsInstaller self) {
		Base(self);

		Zenject.DiContainer container = self.Container;
		SettingsManager settingsManager = Injected<SettingsManager>.Resolve(container)!;
		string hostname = settingsManager.settings.customServer.hostName?.ToLower() ?? string.Empty;
		int port = networkConfigSO.masterServerPort;
		if(hostname.Contains(":") == true) {
			int.TryParse(hostname.Split(':')[1], out port);
			hostname = hostname.Split(':')[0];
		}
		container.Rebind<INetworkConfig>().FromInstance(new CustomNetworkConfig(networkConfigSO, hostname, port, true)).AsSingle();
	}
}

File: BeatUpClient/src/Patches/1_MainSystemInit.cpp
#include "../global.hpp"
#include "../Injected.hpp"
#include <GlobalNamespace/BeatmapCharacteristicCollectionSO.hpp>
#include <GlobalNamespace/BeatmapLevelsModel.hpp>
#include <GlobalNamespace/BoolSO.hpp>
#include <GlobalNamespace/CustomLevelLoader.hpp>
#include <GlobalNamespace/CustomNetworkConfig.hpp>
#include <GlobalNamespace/MainSettingsModelSO.hpp>
#include <GlobalNamespace/MainSystemInit.hpp>
#include <GlobalNamespace/MultiplayerSessionManager.hpp>
#include <GlobalNamespace/MultiplayerStatusModel.hpp>
#include <GlobalNamespace/QuickPlaySetupModel.hpp>
using namespace BeatUpClient;

static bool restoreDefault = false;
BSC_MAKE_HOOK_MATCH(GlobalNamespace::MainSettingsModelSO::Load, void, GlobalNamespace::MainSettingsModelSO *self, bool forced) {
	base(self, forced);
	restoreDefault |= !self->useCustomServerEnvironment->value;
	self->useCustomServerEnvironment->set_value(true);
	logger->info("GlobalNamespace::MainSettingsModelSO::Load()");
}

BSC_MAKE_HOOK_MATCH(GlobalNamespace::MainSettingsModelSO::Save, void, GlobalNamespace::MainSettingsModelSO *self) {
	logger->info("GlobalNamespace::MainSettingsModelSO::Save()");
	self->useCustomServerEnvironment->value = !restoreDefault;
	base(self);
	self->useCustomServerEnvironment->value = true;
}

BSC_MAKE_HOOK_MATCH(GlobalNamespace::MainSystemInit::InstallBindings, void, GlobalNamespace::MainSystemInit *self, Zenject::DiContainer *container) {
	base(self, container);
	// customServerHostName = ____mainSettingsModel.customServerHostName; // TODO: stub
	// NetworkConfigSetup(__instance._networkConfig); // TODO: stub
	Injected<GlobalNamespace::BeatmapCharacteristicCollectionSO>::Resolve<GlobalNamespace::BeatmapCharacteristicCollectionSO>(container);
	Injected<GlobalNamespace::BeatmapLevelsModel>::Resolve<GlobalNamespace::BeatmapLevelsModel>(container);
	Injected<GlobalNamespace::CustomLevelLoader>::Resolve<GlobalNamespace::CustomLevelLoader>(container);
	Injected<GlobalNamespace::CustomNetworkConfig>::Resolve<GlobalNamespace::INetworkConfig>(container);
	Injected<GlobalNamespace::IMultiplayerStatusModel>::Resolve<GlobalNamespace::IMultiplayerStatusModel>(container);
	Injected<GlobalNamespace::IQuickPlaySetupModel>::Resolve<GlobalNamespace::IQuickPlaySetupModel>(container);
	Injected<GlobalNamespace::MultiplayerSessionManager>::Resolve<GlobalNamespace::IMultiplayerSessionManager>(container)->SetLocalPlayerState("modded", true);
	// Net.Setup(Injected<GlobalNamespace::IMenuRpcManager>::Resolve<GlobalNamespace::IMenuRpcManager>(container)); // TODO: stub
}
